File: ios/LndGrpcLightning.swift
//
//  LndGrpcLightning.swift
//  Core2
//
//  Created by ueno on 2023/04/15.
//

import Foundation
import Lndmobile
import GRPC
import NIO
import NIOSSL
import NIOHPACK

@objc(LndGrpcLightning)
class LndGrpcLightning : RCTEventEmitter {
  var subscribeBackupsNum = 0

  @objc
  override static func requiresMainQueueSetup() -> Bool {
    return false
  }

  @objc
  override func supportedEvents() -> [String]! {
    return [
      "grpc_watchLnInv",
      "grpc_watchTx",
      "grpc_watchCh",
      "grpc_watchChanBackup"
    ]
  }

  @objc
  func startWatchLnInvoices(
    _ startDate: Double,
    adminMacaroon: String,
    resolver resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    let eventName = "grpc_watchLnInv"

    do {
      let headers: HPACKHeaders = ["macaroon": adminMacaroon]
      let channel = try LndGrpc.getInstance().createChannel()
      func callback(invoice: Lnrpc_Invoice) {
        if (invoice.state.rawValue == Lnrpc_InvoiceHTLCState.settled.rawValue && invoice.settleDate >= Int64(startDate)) {
          let rhash = dataToHex(invoice.rHash)
          let result: [String: Any] = [
            "event": "settled",
            "param": rhash
          ]
          sendEvent(withName: eventName, body: result)
        }
      }
      func completeHanlder() {
        let result: [String: Any] = [
          "event": "completed",
        ]
        sendEvent(withName: eventName, body: result)
      }
      func errorHandler(_ error: String?) {
        let result: [String: Any] = [
          "event": "error",
          "param": error ?? "unknown error"
        ]
        sendEvent(withName: eventName, body: result)
      }
      let client = Lnrpc_LightningNIOClient(channel: channel)
      let req = Lnrpc_InvoiceSubscription()
      let callOption = CallOptions(customMetadata: headers) // timeoutはcallbackも含む？
      let call = client.subscribeInvoices(req, callOptions: callOption, handler: callback)
      call.status.whenComplete { response in
        print("startWatchLnInvoices complete: \(response)")
        switch (response) {
          case .success(let status):
            if (status.code.rawValue == 0) {
              completeHanlder()
            } else {
              errorHandler(status.message)
            }
            break
          default:
            errorHandler("unknown error")
            break
        }
      }
      resolve(nil)
    } catch {
      print("startWatchLnInvoices error: \(error)")
      reject("LndGrpcLightning", "startWatchLnInvoices", error)
    }
  }
  
  @objc
  func closeChannelStart(
    _ fundingTxid: String,
    outPointIndex: Double,
    forceClose: Bool,
    satPerVByte: Double,
    adminMacaroon: String,
    resolver resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    do {
      var done = false
      let headers: HPACKHeaders = ["macaroon": adminMacaroon]
      let channel = try LndGrpc.getInstance().createChannel()
      func callback(update: Lnrpc_CloseStatusUpdate) {
        print("closeChannelStart update: \(update)")
        if (!done) {
          done = true
          resolve(true)
        }
      }
      let client = Lnrpc_LightningNIOClient(channel: channel)
      var channelPoint = Lnrpc_ChannelPoint()
      channelPoint.fundingTxidStr = fundingTxid
      channelPoint.outputIndex = UInt32(outPointIndex)
      var req = Lnrpc_CloseChannelRequest()
      req.channelPoint = channelPoint
      if (forceClose) {
        req.force = true
      } else {
        req.satPerVbyte = UInt64(satPerVByte)
      }
      let callOption = CallOptions(customMetadata: headers)
      let call = client.closeChannel(req, callOptions: callOption, handler: callback)
      call.status.whenComplete { response in
        print("closeChannelStart complete: \(response)")
        if (!done) {
          done = true
          resolve(true)
        }
      }
    } catch {
      print("closeChannelStart error: \(error)")
      resolve(false)
    }
  }

  @objc
  func subscribeTransactions(
    _ adminMacaroon: String,
    resolver resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    let eventName = "grpc_watchTx"

    do {
      let headers: HPACKHeaders = ["macaroon": adminMacaroon]
      let channel = try LndGrpc.getInstance().createChannel()
      func callback(transaction: Lnrpc_Transaction) {
        if (transaction.numConfirmations > 0) {
          print("subscribeTransactions.callback: \(transaction)")
          let result: [String: Any] = [
            "event": "detect",
            "txid": transaction.txHash,
            "confirm": transaction.numConfirmations,
            "amount": transaction.amount,
            "height": transaction.blockHeight,
            "addresses": transaction.destAddresses
          ]
          sendEvent(withName: eventName, body: result)
        }
      }
      func completeHanlder() {
        let result: [String: Any] = [
          "event": "completed",
        ]
        sendEvent(withName: eventName, body: result)
      }
      func errorHandler(_ error: String?) {
        let result: [String: Any] = [
          "event": "error",
          "reason": error ?? "unknown error"
        ]
        sendEvent(withName: eventName, body: result)
      }
      let client = Lnrpc_LightningNIOClient(channel: channel)
      let req = Lnrpc_GetTransactionsRequest()
      let callOption = CallOptions(customMetadata: headers)
      let call = client.subscribeTransactions(req, callOptions: callOption, handler: callback)
      call.status.whenComplete { response in
        print("subscribeTransactions complete: \(response)")
        switch (response) {
          case .success(let status):
            if (status.code.rawValue == 0) {
              completeHanlder()
            } else {
              errorHandler(status.message)
            }
            break
          default:
            errorHandler("unknown error")
            break
        }
      }
      resolve(nil)
    } catch {
      print("subscribeTransactions error: \(error)")
      reject("LndGrpcLightning", "subscribeTransactions", error)
    }
  }

  @objc
  func subscribeChannelEvents(
    _ adminMacaroon: String,
    resolver resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    let eventName = "grpc_watchCh"

    do {
      let headers: HPACKHeaders = ["macaroon": adminMacaroon]
      let channel = try LndGrpc.getInstance().createChannel()
      func callback(event: Lnrpc_ChannelEventUpdate) {
        print("subscribeChannelEvents.callback: \(event)")
        switch event.type {
        case .openChannel:
          let result: [String: Any] = [
            "event": "OPEN_CHANNEL",
            "channel_point": event.openChannel.channelPoint
          ]
          sendEvent(withName: eventName, body: result)
        case .closedChannel:
          let result: [String: Any] = [
            "event": "CLOSED_CHANNEL",
            "channel_point": event.closedChannel.channelPoint
          ]
          sendEvent(withName: eventName, body: result)
        default:
          print("subscribeChannelEvents.callback: \(event.type)")
          break
        }
      }
      func completeHanlder() {
        let result: [String: Any] = [
          "event": "completed",
        ]
        sendEvent(withName: eventName, body: result)
      }
      func errorHandler(_ error: String?) {
        let result: [String: Any] = [
          "event": "error",
          "reason": error ?? "unknown error"
        ]
        sendEvent(withName: eventName, body: result)
      }
      let client = Lnrpc_LightningNIOClient(channel: channel)
      let req = Lnrpc_ChannelEventSubscription()
      let callOption = CallOptions(customMetadata: headers)
      let call = client.subscribeChannelEvents(req, callOptions: callOption, handler: callback)
      call.status.whenComplete { response in
        print("subscribeChannelEvents complete: \(response)")
        switch (response) {
          case .success(let status):
            if (status.code.rawValue == 0) {
              completeHanlder()
            } else {
              errorHandler(status.message)
            }
            break
          default:
            errorHandler("unknown error")
            break
        }
      }
      resolve(nil)
    } catch {
      print("subscribeChannelEvents error: \(error)")
      reject("LndGrpcLightning", "subscribeChannelEvents", error)
    }
  }

  @objc
  func subscribeChannelBackups(
    _ adminMacaroon: String,
    resolver resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    let eventName = "grpc_watchChanBackup"

    do {
      let headers: HPACKHeaders = ["macaroon": adminMacaroon]
      let channel = try LndGrpc.getInstance().createChannel()
      func callback(backup: Lnrpc_ChanBackupSnapshot) {
        print("subscribeChannelBackups.callback: \(backup)")
        let backupBase64 = backup.multiChanBackup.multiChanBackup.base64EncodedString()
        var chanPoints: [[String: Any]] = []
        for chanBackup in backup.singleChanBackups.chanBackups {
          let chanPoint: [String: Any] = [
            "funding_txid_str": chanBackup.chanPoint.fundingTxidStr,
            "output_index": chanBackup.chanPoint.outputIndex
          ]
          chanPoints.append(chanPoint)
        }
        subscribeBackupsNum += 1
        let result: [String: Any] = [
          "event": "detect",
          "count": subscribeBackupsNum,
          "backupBase64": backupBase64,
          "chanPoints": chanPoints
        ]
        sendEvent(withName: eventName, body: result)
      }
      func completeHanlder() {
        let result: [String: Any] = [
          "event": "completed",
        ]
        sendEvent(withName: eventName, body: result)
      }
      func errorHandler(_ error: String?) {
        let result: [String: Any] = [
          "event": "error",
          "reason": error ?? "unknown error"
        ]
        sendEvent(withName: eventName, body: result)
      }
      let client = Lnrpc_LightningNIOClient(channel: channel)
      let req = Lnrpc_ChannelBackupSubscription()
      let callOption = CallOptions(customMetadata: headers)
      let call = client.subscribeChannelBackups(req, callOptions: callOption, handler: callback)
      call.status.whenComplete { response in
        print("subscribeChannelBackups complete: \(response)")
        switch (response) {
          case .success(let status):
            if (status.code.rawValue == 0) {
              completeHanlder()
            } else {
              errorHandler(status.message)
            }
            break
          default:
            errorHandler("unknown error")
            break
        }
      }
      resolve(nil)
    } catch {
      print("subscribeChannelBackups error: \(error)")
      reject("LndGrpcLightning", "subscribeChannelBackups", error)
    }
  }
}

File: android/app/src/main/java/com/nayuta/core2/LndGrpcLightning.kt
package com.nayuta.core2

import android.util.Base64
import android.util.Log
import com.facebook.react.bridge.*
import com.nayuta.core2.LndGrpc.grpcAndroidChannel
import com.nayuta.core2.LndGrpc.sendEventToJs
import io.grpc.stub.StreamObserver
import java.nio.charset.Charset
import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit

// lightning.proto
//  android/app/build/generated/source/proto/xxx/grpc/lnrpc/LightningGrpc.java
class LndGrpcLightning(private val reactContext: ReactApplicationContext, private val lndController: LndReactController) : ReactContextBaseJavaModule(reactContext) {

    override fun getName(): String {
        return "LndGrpcLightning"
    }

    @ReactMethod
    fun addListener(eventName: String?) {
        // Keep: Required for RN built in Event Emitter Calls.
    }

    @ReactMethod
    fun removeListeners(count: Double) {
        // Keep: Required for RN built in Event Emitter Calls.
    }

    private var isLnInvoiceWatching = false
    private var isSubscribeTransactions = false
    private var isSubscribeChannelEvents = false
    private var isSubscribeBackups = false
    private var subscribeBackupsNum = 0
    fun ByteArray.toHexString() = joinToString("") { "%02x".format(it) }

    /** notify settled invoice
     *
     * @param startDate notify invoices.settle_date >= startDate
     * @param adminMacaroon admin macaroon
     * @param promise promise
     */
    @ReactMethod
    fun startWatchLnInvoices(startDate: Int, adminMacaroon: String, promise: Promise) {
        if (isLnInvoiceWatching) {
            Log.i(TAG, "startWatchLnInvoices() already watching")
            promise.resolve(null)
            return
        }
        isLnInvoiceWatching = true

        Log.i(TAG, "startWatchLnInvoices(${startDate})")

        // https://grpc.io/docs/platforms/android/java/basics/#client-side-streaming-rpc
        val responseObserver: StreamObserver<lnrpc.LightningOuterClass.Invoice> = object :
            StreamObserver<lnrpc.LightningOuterClass.Invoice> {
            val TAG = "startWatchLnInvoices"
            val eventName = "grpc_watchLnInv"

            override fun onNext(summary: lnrpc.LightningOuterClass.Invoice) {
                Log.i(TAG, "onNext: ${summary.getSettleDate()}")
                if ((summary.getState() == lnrpc.LightningOuterClass.Invoice.InvoiceState.SETTLED) && (summary.getSettleDate() >= startDate.toLong())) {
                    val result = Arguments.createMap()
                    result.putString("event", "settled")
                    result.putString("param", summary.getRHash().toByteArray().toHexString())
                    sendEventToJs(reactContext, eventName, result)
                }
            }

            override fun onError(t: Throwable) {
                Log.e(TAG, "onError: $t")
                isLnInvoiceWatching = false
                val result = Arguments.createMap()
                result.putString("event", "error")
                result.putString("param", t.toString())
                sendEventToJs(reactContext, eventName, result)
            }

            override fun onCompleted() {
                Log.e(TAG, "onCompleted")
                isLnInvoiceWatching = false
                val result = Arguments.createMap()
                result.putString("event", "completed")
                sendEventToJs(reactContext, eventName, result)
            }
        }

        val req = lnrpc.LightningOuterClass.InvoiceSubscription.newBuilder()
            .build()
        val asyncStub = lnrpc.LightningGrpc.newStub(grpcAndroidChannel(reactContext, lndController)).withCallCredentials(
            LndGrpc.MacaroonCallCredential(adminMacaroon)
        )
        asyncStub.subscribeInvoices(req, responseObserver)

        Log.i(TAG, "startWatchLnInvoices() - started")
        promise.resolve(null)
    }

    /** close channel
     *      https://api.lightning.community/#closechannel
     *
     * @param adminMacaroon admin macaroon
     */
    @ReactMethod
    fun closeChannelStart(fundingTxid: String, outPointIndex: Int, forceClose: Boolean, satPerVByte: Int, adminMacaroon: String, promise: Promise) {
        Log.i(TAG, "closeChannelStart(${fundingTxid}:${outPointIndex} force=${forceClose} satPerVbyte=${satPerVByte})")

        var closeStarted = false
        val finishLatch = CountDownLatch(1)
        val closeChannelResponseObserver: StreamObserver<lnrpc.LightningOuterClass.CloseStatusUpdate> = object :
            StreamObserver<lnrpc.LightningOuterClass.CloseStatusUpdate> {
            val TAG = "closeChannelStart"

            override fun onNext(summary: lnrpc.LightningOuterClass.CloseStatusUpdate) {
                Log.i(TAG, "onNext: hasClosePending=${summary.hasClosePending()}, hasChanClose=${summary.hasChanClose()}")
                closeStarted = true
                finishLatch.countDown()
            }

            override fun onError(t: Throwable) {
                Log.e(TAG, "onError: $t")
                finishLatch.countDown()
            }

            override fun onCompleted() {
                Log.e(TAG, "onCompleted")
                closeStarted = true
                finishLatch.countDown()
            }
        }

        // gRPC Request: lnrpc.CloseChannelRequest
        //  android/app/build/generated/source/proto/xxx/java/LightningOuterClass.java
        val channelPoint = lnrpc.LightningOuterClass.ChannelPoint.newBuilder()
            .setFundingTxidStr(fundingTxid)
            .setOutputIndex(outPointIndex)
            .build()
        val req = lnrpc.LightningOuterClass.CloseChannelRequest.newBuilder()
            .setChannelPoint(channelPoint)
            .setForce(forceClose)
            .setSatPerVbyte(if (forceClose) 0 else satPerVByte.toLong()) // force時に0以外を設定すると"force closing a channel uses a pre-defined fee"になる
            .build()
        val asyncStub = lnrpc.LightningGrpc.newStub(grpcAndroidChannel(reactContext, lndController)).withCallCredentials(
            LndGrpc.MacaroonCallCredential(adminMacaroon)
        )
        asyncStub.closeChannel(req, closeChannelResponseObserver)

        if (!finishLatch.await(1, TimeUnit.MINUTES)) {
            Log.e(TAG, "closeChannelStart() - timeout")
        }

        Log.i(TAG, "closeChannelStart() - started=$closeStarted")
        promise.resolve(closeStarted)
    }

    /** subscribe transactions
     *      https://api.lightning.community/#subscribetransactions
     *
     * @param adminMacaroon admin macaroon
     */
    @ReactMethod
    fun subscribeTransactions(adminMacaroon: String, promise: Promise) {
        Log.i(TAG, "subscribeTransactions()")

        if (isSubscribeTransactions) {
            Log.i(TAG, "subscribeTransactions() already watching")
            promise.resolve(null)
            return
        }
        isSubscribeTransactions = true

        // https://grpc.io/docs/platforms/android/java/basics/#client-side-streaming-rpc
        val responseObserver: StreamObserver<lnrpc.LightningOuterClass.Transaction> = object :
            StreamObserver<lnrpc.LightningOuterClass.Transaction> {
            val TAG = "subscribeTransactions"
            val eventName = "grpc_watchTx"

            override fun onNext(transaction: lnrpc.LightningOuterClass.Transaction) {
                Log.i(TAG, "onNext: ${transaction.getTxHash()}")
                if ((transaction.getNumConfirmations() > 0)) {
                    val result = Arguments.createMap()
                    result.putString("event", "detect")
                    result.putString("txid", transaction.getTxHash())
                    result.putInt("confirm", transaction.getNumConfirmations())
                    result.putInt("amount", transaction.getAmount().toInt())
                    result.putInt("height", transaction.getBlockHeight())
                    val addresses = WritableNativeArray()
                    for (addr in transaction.getDestAddressesList()) {
                        Log.d(TAG, addr)
                        addresses.pushString(addr)
                    }
                    result.putArray("addresses", addresses)
                    sendEventToJs(reactContext, eventName, result)
                }
            }

            override fun onError(t: Throwable) {
                Log.e(TAG, "onError: $t")
                isSubscribeTransactions = false
                val result = Arguments.createMap()
                result.putString("event", "error")
                result.putString("reason", t.toString())
                sendEventToJs(reactContext, eventName, result)
            }

            override fun onCompleted() {
                Log.e(TAG, "onCompleted")
                isSubscribeTransactions = false
                val result = Arguments.createMap()
                result.putString("event", "completed")
                sendEventToJs(reactContext, eventName, result)
            }
        }

        val req = lnrpc.LightningOuterClass.GetTransactionsRequest.newBuilder()
            .build()
        val asyncStub = lnrpc.LightningGrpc.newStub(grpcAndroidChannel(reactContext, lndController)).withCallCredentials(
            LndGrpc.MacaroonCallCredential(adminMacaroon)
        )
        asyncStub.subscribeTransactions(req, responseObserver)

        Log.i(TAG, "startWatchLnInvoices() - started")
        promise.resolve(null)
    }


    /** subscribe channel events
     *      https://api.lightning.community/#subscribechannelevents
     *
     * @param adminMacaroon admin macaroon
     */
    @ReactMethod
    fun subscribeChannelEvents(adminMacaroon: String, promise: Promise) {
        Log.i(TAG, "subscribeChannelEvents()")

        if (isSubscribeChannelEvents) {
            Log.i(TAG, "subscribeChannelEvents() already watching")
            promise.resolve(null)
            return
        }
        isSubscribeChannelEvents = true

        // https://grpc.io/docs/platforms/android/java/basics/#client-side-streaming-rpc
        val responseObserver: StreamObserver<lnrpc.LightningOuterClass.ChannelEventUpdate> = object :
            StreamObserver<lnrpc.LightningOuterClass.ChannelEventUpdate> {
            val TAG = "subscribeChannelEvents"
            val eventName = "grpc_watchCh"

            override fun onNext(event: lnrpc.LightningOuterClass.ChannelEventUpdate) {
                Log.i(TAG, "onNext: eventupdate: ${event.getType()}")
                when (event.getType()) {
                    lnrpc.LightningOuterClass.ChannelEventUpdate.UpdateType.OPEN_CHANNEL -> {
                        val result = Arguments.createMap()
                        result.putString("event", "OPEN_CHANNEL")
                        result.putString("channel_point", event.getOpenChannel().getChannelPoint())
                        sendEventToJs(reactContext, eventName, result)
                    }
                    lnrpc.LightningOuterClass.ChannelEventUpdate.UpdateType.CLOSED_CHANNEL -> {
                        val result = Arguments.createMap()
                        result.putString("event", "CLOSED_CHANNEL")
                        result.putString("channel_point", event.getClosedChannel().getChannelPoint())
                        sendEventToJs(reactContext, eventName, result)
                    }
                }
            }

            override fun onError(t: Throwable) {
                Log.e(TAG, "onError: $t")
                isSubscribeChannelEvents = false
                val result = Arguments.createMap()
                result.putString("event", "error")
                result.putString("reason", t.toString())
                sendEventToJs(reactContext, eventName, result)
            }

            override fun onCompleted() {
                Log.e(TAG, "onCompleted")
                isSubscribeChannelEvents = false
                val result = Arguments.createMap()
                result.putString("event", "completed")
                sendEventToJs(reactContext, eventName, result)
            }
        }

        val req = lnrpc.LightningOuterClass.ChannelEventSubscription.newBuilder()
            .build()
        val asyncStub = lnrpc.LightningGrpc.newStub(grpcAndroidChannel(reactContext, lndController)).withCallCredentials(
            LndGrpc.MacaroonCallCredential(adminMacaroon)
        )
        asyncStub.subscribeChannelEvents(req, responseObserver)

        Log.i(TAG, "startWatchLnInvoices() - started")
        promise.resolve(null)
    }

    /** subscribe channel backups
     *      https://api.lightning.community/#subscribechannelbackups
     *
     * @param adminMacaroon admin macaroon
     */
    @ReactMethod
    fun subscribeChannelBackups(adminMacaroon: String, promise: Promise) {
        Log.i(TAG, "subscribeChannelBackups()")

        if (isSubscribeBackups) {
            Log.i(TAG, "subscribeChannelBackups() already watching")
            promise.resolve(null)
            return
        }
        isSubscribeBackups = true

        // https://grpc.io/docs/platforms/android/java/basics/#client-side-streaming-rpc
        val responseObserver: StreamObserver<lnrpc.LightningOuterClass.ChanBackupSnapshot> = object :
            StreamObserver<lnrpc.LightningOuterClass.ChanBackupSnapshot> {
            val TAG = "subscribeChannelBackups"
            val eventName = "grpc_watchChanBackup"

            override fun onNext(backup: lnrpc.LightningOuterClass.ChanBackupSnapshot) {
                Log.i(TAG, "onNext: ${backup.getSingleChanBackups().getChanBackupsCount()}")
                val backupBase64 = Base64.encode(backup.getMultiChanBackup().getMultiChanBackup().toByteArray(), Base64.URL_SAFE or Base64.NO_PADDING or Base64.NO_WRAP)
                val result = Arguments.createMap()
                result.putString("event", "detect")
                result.putInt("count", ++subscribeBackupsNum)
                result.putString(
                    "backupBase64", backupBase64!!.toString(Charset.forName("US-ASCII"))
                )
                val chanPoints = Arguments.createArray()
                for (chanBackup in backup.getSingleChanBackups().getChanBackupsList()) {
                    val chanPoint = Arguments.createMap()
                    chanPoint.putString("funding_txid_str", chanBackup.getChanPoint().getFundingTxidBytes().toByteArray().reversedArray().toHexString())
                    chanPoint.putInt("output_index", chanBackup.getChanPoint().getOutputIndex())
                    chanPoints.pushMap(chanPoint)
                }
                result.putArray("chanPoints", chanPoints)
                sendEventToJs(reactContext, eventName, result)
            }

            override fun onError(t: Throwable) {
                Log.e(TAG, "onError: $t")
                isSubscribeTransactions = false
                val result = Arguments.createMap()
                result.putString("event", "error")
                result.putString("reason", t.toString())
                sendEventToJs(reactContext, eventName, result)
            }

            override fun onCompleted() {
                Log.e(TAG, "onCompleted")
                isSubscribeTransactions = false
                val result = Arguments.createMap()
                result.putString("event", "completed")
                sendEventToJs(reactContext, eventName, result)
            }
        }

        val req = lnrpc.LightningOuterClass.ChannelBackupSubscription.newBuilder()
            .build()
        val asyncStub = lnrpc.LightningGrpc.newStub(grpcAndroidChannel(reactContext, lndController)).withCallCredentials(
            LndGrpc.MacaroonCallCredential(adminMacaroon)
        )
        asyncStub.subscribeChannelBackups(req, responseObserver)

        Log.i(TAG, "startWatchLnInvoices() - started")
        promise.resolve(null)
    }

    companion object {
        private val TAG = LndGrpcLightning::class.java.name
    }
}